### PR TITLE
Assembler: Add flag pattern-assembler/skip-rendered-pattern-cache in Calypso

### PR DESCRIPTION
**Pull request by miksansegundo**
Related to D131807-code

## Proposed Changes

* Add flag pattern-assembler/skip-rendered-pattern-cache in Calypso
* Set `staleTime: 0` when the flag is enabled.

> [!NOTE]  
> Initially, I thought that we could use the flag `pattern-assembler/v2` to skip the cache. It sounds better to use a new flag so we can enable this feature any time passing the flag in the URL.

> [!WARNING]  
> This PR alone is not enough to skip the cache because the request to fetch patterns from the PTK API `/rest/v1.1/ptk/patterns/en` is also cached. See D131807-code

## Testing Instructions

* Using the Calypso Live link below
* Access `/setup/with-theme-assembler/patternAssembler?siteSlug={ SITE }`
* Open "Chrome dev tools > Network" and verify that the renderer requests `/wpcom/v2/sites/215354566/block-renderer/patterns/render` are done whenever you refresh the browser. 

https://github.com/Automattic/wp-calypso/assets/1881481/44bafdd4-29dc-46f5-b9c2-4a1bd096464d

## Pre-merge Checklist

- [X] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [X] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [X] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [X] Have you checked for TypeScript, React or other console errors?
- [X] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [X] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [X] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: packages/block-renderer/src/hooks/use-rendered-patterns.ts [hunk 1]
```diff
@@ -41,7 +42,7 @@ const useRenderedPatterns = (
 	const queries = Object.entries( patternIdsByCategory ).map( ( [ category, patternIds ] ) => ( {
 		queryKey: [ 'rendered-patterns', siteId, stylesheet, category, patternIds, siteInfo ],
 		queryFn: () => fetchRenderedPatterns( siteId, stylesheet, category, patternIds, siteInfo ),
-		staleTime: 1000 * 60 * 5, // 5 minutes
+		staleTime: isEnabled( 'pattern-assembler/skip-rendered-pattern-cache' ) ? 0 : 1000 * 60 * 5, // 5 minutes
```
Comment (arthur791004) [inline packages/block-renderer/src/hooks/use-rendered-patterns.ts:45]: [nit] We can simply call it `pattern-assembler/disable-cache` and then we can apply the same flag to other places if needed

**Review by arthur791004 — APPROVED** [review]

**Comment by fushar** [comment]
OK, I did some research, and I think we got this wrong.

During the lifetime of the pattern assembler React components, patterns are very rarely modified. So, `staleTime` should be practically `Infinity`, not 5 minutes or 0.

However, when we refresh the browser, we should be re-requesting the patterns. If there's cache, ideally it should be on the backend side. Note that during a regular usage of Assembler flow, a user won't be refreshing the browser that often, so we should be fine.

```diff
diff --git a/packages/block-renderer/src/hooks/use-rendered-patterns.ts b/packages/block-renderer/src/hooks/use-rendered-patterns.ts
index f3a7e40117f..b1e84248ab4 100644
--- a/packages/block-renderer/src/hooks/use-rendered-patterns.ts
+++ b/packages/block-renderer/src/hooks/use-rendered-patterns.ts
@@ -41,8 +41,11 @@ const useRenderedPatterns = (
 	const queries = Object.entries( patternIdsByCategory ).map( ( [ category, patternIds ] ) => ( {
 		queryKey: [ 'rendered-patterns', siteId, stylesheet, category, patternIds, siteInfo ],
 		queryFn: () => fetchRenderedPatterns( siteId, stylesheet, category, patternIds, siteInfo ),
-		staleTime: 1000 * 60 * 5, // 5 minutes
+		staleTime: Infinity,
 		refetchOnWindowFocus: false,
+		meta: {
+			persist: false,
+		},
 	} ) );
```

This way, this will be (and SHOULD be) the same as how we're requesting the initial set of patterns from PTK:

https://github.com/Automattic/wp-calypso/blob/310a632b8de8bcaa47219a4df2112a6f3584a90d/client/landing/stepper/declarative-flow/internals/steps-repository/pattern-assembler/hooks/use-dotcom-patterns.ts#L34-L37

In short, I think:

- We shouldn't introduce any new flag in Calypso side.
- We apply the above patch.
- (Optional) Add the "skip cache" logic for PTK endpoint. But I think this is not really necessary, since we will be updating patterns the most, not adding / deleting from the set of patterns. So I think it's fine if we need to wait for 5 minutes after adding new patterns.
- (Optional) Add caching logic for `/block-renderer` endpoint; but I believe this won't help much since (I think) it's the HTML rendering (css, images etc) that takes the most "lag", not fetching the HTML data from the backend (I don't have any data though)

What do you think?

**Comment by miksansegundo** [comment]
@fushar sounds good to me, nice plan, thanks.
I'll wait to hear @arthur791004 thoughts on the plan

**Comment by arthur791004** [comment]
> However, when we refresh the browser, we should be re-requesting the patterns. If there's cache, ideally it should be on the backend side.

But it means we have to re-fetch rendered patterns after the checkout 🥲

**Comment by fushar** [comment]
> But it means we have to re-fetch rendered patterns after the checkout 🥲

Yeah that's included in my statement above 🙈 :

> Note that during a regular usage of Assembler flow, a user won't be refreshing the browser that often

**Comment by miksansegundo** [comment]
We still have to address https://github.com/Automattic/wp-calypso/issues/85214 by not fetching patterns again on that endpoint or by adding a server-side cache.
